### DISCUSSION TITLE
[ARCHIVED] Поговорим об импортах...

### DISCUSSION THREAD

**Discussion by azinit** [feature-sliced/documentation#52] [Archive]
## :fire: Не нужны нам ваши импорты!

Среди некоторых разработчиков все чаще встречаю заявления из разряда:

- *"**Уже несколько месяцев живу со свернутой шапкой импортов**"*
- *"Никогда не читаю импорты. Студия их заботливо скрывает чтобы не мешали код читать. Неиспользуемые если есть - подскажет. Потому я не особенно понимаю зачем сортировать и ещё с этим возиться."*
- *"Не забочусь об импортах, автоматически подтягиваются"*
- *"**А зачем сортировать импорты? А что это даёт? Ускоряет разработку, позволяет делать меньше багов?**"*

## :fire: Но есть и другой лагерь
- *"Есть особое эстетическое удовольствие, когда импорты расположены в нужном порядке"*
- *"Импорты часто вызывают холивары при ревью, и код-ревьюверы порой заваливают этим правки"* - **а потому их линтование и автоформатирование, помогает этого избежать**
- *"Автоимпорты подходят, только когда ты явно знаешь, что ты хочешь импортить (автокомплит же при ручном указании импортов - в этом помогает)"*
- *"Также, человек со скрытыми импортами - не видит внутренних ошибок линтера, например касаемо того - "какие модули можно импортить, а какие нет", **и если мы завязываем глаза в этом моменте, то недалеко так дойти снова до импорта "кишков" фич, а не его Public API**"*
- *"Плюс к предыдущему пункту - многие сущности могут называться одинаково в модулях (initialState, actions, selectors, effects, ...) - и где гарантии, что при автоимпорте заимпортиться именно то? **Где в принципе гарантии, что автоимпорт будет работать нормально, если мы говорим про серьезное приложение**, где "нельзя импортить все откуда угодно"?"*
- *"**Чаще всего гораздо практичнее сразу писать импорты вручную, т.к. так**им образом в том числе мы можем обнаружить Public API модулей более явно, особенно когда приходим в большой проект, в котором мало знаем что экспортиться (как вы будете импортить "не знаю-что"-"но знаю откуда" автокомплитом?)"*

---

> Не особо велика вероятность, что разговор об импортах попадет в методологию, но лично мне кажется это немаловажным аспектом - **т.к. линтование и разумное использование импортов - один из залогов грамотного использования абстракций из методологии**

**Comment by azinit** [comment]
## В каком вы лагере?

**Reply by azinit** [reply]
- Не вижу смысла вручную задавать импорты
- Уже давно не смотрю/не работаю с шапкой импортов
- Автоимпорты наше все

**Reply by azinit** [reply]
- Предпочитаю быть уверенным в том, что я импортирую
- Для меня бывает полезно, если работа с импортами линтуется и форматируется
- Не всегда/Никогда не доверяю автоимпортам

**Comment by azinit** [comment]
Если говорить обо мне - то наверное видно по риторике, что я скорее за **"ручные импорты", т.к. у меня больше контроля при таком подходе**

**Reply by azinit** [reply]
Возвращаясь к методологии, я вижу еще следующую проблему "нормально ли использовать всегда относительные импорты? Абсолютные нас спасут?"

**Reply by azinit** [reply]
Лично я стараюсь придерживаться следующего:
- использовать **абсолютные импорты** для основных и верхнеуровневых абстракций
    > `import { skinsModels } from "entities/skins"`
    >
    > у нас ведь features/pages/processes/entities/shared/app - основные абстракции, и к ним должен быть всегда доступ "свыше" - т.е. привет абсолютные импорты
    

- использовать **относительные импорты** только для ресурсов - которые инкапсулированы ВНУТРИ текущей абстракции 
     > `import { actions } from "../../../models"`



> Как по мне эта проблема коррелирует еще с "Public API" - #41

**Reply by Lonli-Lokli** [reply]
Мое мнение - если я лезу внутрь (`'./'`) - это моя зона ответственности, если я лезу наружу (`'../'`) - чужая.
ПС Брать чужое нехорошо

**Reply by Affiction** [reply]
@Lonli-Lokli 

```
pages/
    /parent-route
        /child-route // models - store for current page
            /grand-child-route-1   // import { modelsFromChildRoute } from '../models';
            /grand-child-route-2   // import { modelsFromChildRoute } from '../models';
            /grand-child-route-3   // import { modelsFromChildRoute } from '../models';
```

В этом случае нормально лезть наружу

**Comment by ilyaagarkov** [comment]
Иногда импорты все равно просматриваю, чтобы понять от чего зависит данный модуль.
В целом работа с импортами это рутина, так что должна быть максимально автоматизирована. Думаю два лагеря будут довольны, если линтер будет сортировать импорты сам

**Reply by azinit** [reply]
> В целом работа с импортами это рутина, так что должна быть максимально автоматизирована. Думаю два лагеря будут довольны, если линтер будет сортировать импорты сам

Да, я как раз выше и писал, что *т.к. мы хотим больше контроля, но меньше рутины*  - **линтование и форматироание для импортов - маст-хев**

> Привет, [`import/order`](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/order.md), [`sort-imports`](https://eslint.org/docs/4.0.0/rules/sort-imports), [`eslint-plugin-boundaries`](https://github.com/javierbrea/eslint-plugin-boundaries)

**Comment by azinit** [comment]
@zarabotaet приглашаю к дискуссии))

**Comment by zarabotaet** [comment]
![image](https://user-images.githubusercontent.com/15930980/111042481-4aa77e00-8446-11eb-9968-0e88547a2f30.png)
Так выглядят файлы для меня 95% времени разработки 

![image](https://user-images.githubusercontent.com/15930980/111042546-89d5cf00-8446-11eb-8a0a-0e1900fb0331.png)
Вот так выглядит секция импортов

![image](https://user-images.githubusercontent.com/15930980/111042591-bdb0f480-8446-11eb-88ea-41ae2f54d0a1.png)
![image](https://user-images.githubusercontent.com/15930980/111042596-c5709900-8446-11eb-8c2c-f65f2724a828.png)
![image](https://user-images.githubusercontent.com/15930980/111042636-ffda3600-8446-11eb-9375-56744169b508.png)
Иногда заглядывая делаю сворачиваю в неймспейс пачку импортов средствами vscode

![image](https://user-images.githubusercontent.com/15930980/111042656-1e403180-8447-11eb-8bb1-2d39621ddc3b.png)
Мой конфиг линта для сортировки и автоматического разбиения на группы импортов

**Reply by zarabotaet** [reply]
Делая скрины я даже не заметил что именно на этом проекте есть алиасы, так я перестал замечать их наличие или остуствие всегда))) так что на одну заботу меньше - чуть жизни фронта легче ИМХО
![image](https://user-images.githubusercontent.com/15930980/111042763-9eff2d80-8447-11eb-87da-b55229cf20a8.png)
Прекрасные относительные импорты. К тому же vscode с относительными лучше дружит.

**Reply by azinit** [reply]
Можешь отметиться)
https://github.com/feature-sliced/wiki/discussions/52#discussioncomment-477425

**Reply by azinit** [reply]
Ну окей
- *свернутые импорты*
- *тулкит vscode*
- *конфиг линта для сортировки и разбиения*
- *на одну заботу меньше - чуть жизни фронта легче*

Да, да - звучит ~~хайпово~~ здорово))

1. Но как ты при таком подходе импортишь **согласно Public API и не залезая в кишки модуля**? Как у тебя будет **резолвиться коллизия имен**?
   > Например, внутри фичи `cart` ты захотел импортнуть `actions` из модели/store. И где гарантии - что actions импортнутся откуда нужно? А если ты захотел обратиться к экшенам вне фичи - где гарантии - что автоимпорт предложит тебе заимпортить Public API и на внутренности сразу?
2. Как ты с таким подходом будешь импортить вещи в проекте, **где ты не знаешь "все как свои 5 пальцев"**? Как будут импортить другие ребята, которые придут в твой проект? Они сразу разве заранее будут знать в каких модулях - какой Public API ?
3. На каждом мало-мальски серьезном проекте обычно импорты линтятся согласно определенным правилам (простой пример: "нельзя из фичи импортить другую") - **опять же для того чтобы ты не выстрелил в ногу себе, команде и проекту**
     Возможно ты - опять же - знаешь свой проект прекрасно, но что будет когда ты придешь в другой? (или когда к тебе кто-нибудь придет). **Будешь импортить все что "захочешь", даже если линтер будет подсвечивать "ошибку импортов"?**
     > А ты ведь этого и не увидишь, т.к. у тебя шапка всегда свернута и ты якобы становишься "счастливее" как фронт, т.к. ты просто "закрыл глаза на одну проблему"

**Reply by azinit** [reply]
4. \+ аргумент от @ilyaagarkov : *"Иногда импорты все равно просматриваю, чтобы понять от чего зависит данный модуль"*

**Reply by azinit** [reply]
> @zarabotaet но за плагин [`simple-import-sort`](https://github.com/lydell/eslint-plugin-simple-import-sort) все же спасибо, не находил такого

**Reply by azinit** [reply]
5. **Еще скорее вброс, но все же**: как так вышло, что человек топящий за автоимпорты и легкую жизнь разраба без импортов - не автоимпортит из React хуки, а использует их через неймспейс?))

![image](https://user-images.githubusercontent.com/42924400/111055469-2e780100-8487-11eb-8b2a-4611ea76b15b.png)

**Reply by zarabotaet** [reply]
1. я пользуюсь полуавтоматическим автоимпортом 
![image](https://user-images.githubusercontent.com/15930980/111065475-aca5c900-84c2-11eb-97f8-95de713e45f0.png)
2. Паблик апи любого модуля у меня выделен сразу (если у модуля есть экспортируемые не паблик сущности)
![image](https://user-images.githubusercontent.com/15930980/111065519-de1e9480-84c2-11eb-9b90-c95a8bbe5eaa.png)
3. любые прекоммит хуки решают все разом для всей команды (я использую lefthook). Если у тиммейта появляются вопросы почему что то откуда то нельзя импортировать то вся логика описана в соотвествующем файле md (LEGACY.md  или ARCHITECTURE.md) или вики страничке
4. тоже иногда прсматриваю, madge и аналоги. Не вижу смысла учится строить схемы свзяей в голове.
5. кастомный снипет сразу мне создает экспортируемый FC c импортом React и строкой стейта

**Reply by azinit** [reply]
@zarabotaet спасибо за ответ)

> 1. я пользуюсь полуавтоматическим автоимпортом

Ну я примерно такую картину и представлял, но чет мне кажется мы и так как разрабы мучаемся от большого автокомплита (tabnine, [нейминг-функций](https://github.com/feature-sliced/wiki/discussions/12), и т.п.)

И то что ты привел - конечно ооооччч, было бы здорово, если бы показывало только Public API разрешенный, но скорее всего автокомплит показывает только "все возможные варианты", и среди них найти только Public API... Ну оке, допустим "засчитано" в данном пункте))

*Хотя не особо вижу супер-профит писать полностью путь через автокомплит или же через импорты напрямую*

**Reply by azinit** [reply]
> 2. Паблик апи любого модуля у меня выделен сразу (если у модуля есть экспортируемые не паблик сущности)

В любом случае новым ребятам в команде (или тем кто вернется к проекту позже) - придется "итеративно" путем уточнения импорта получать информацию о модулях, т.е. обычным автокомплитом будет сложно это сделать, т.к. тебе как правило надо "знать что именно должно экспортиться из модуля"

А кроме как "подбором импорта" и "залезания в кишки" - ты этого не узнаешь

**И поскольку мы так топим против второго, то лучше уже "постепенно уточнять импорт", добираясь до Public API нужной вложенности**

**Reply by azinit** [reply]
> 3. любые прекоммит хуки решают все разом для всей команды (я использую lefthook). 

Ага, только человек узнает проблему "не сразу как она появилась", а только на этапе коммита/пуша

И например из-за этого он мог архитектуру построить по фиче совсем не тем путем, каким надо было
Потому своевременная подсветка варнингов в импортах - могла бы быть полезна, если бы "не сворачивали каждый раз шапку импортов"  (**разраб бы сразу увидел проблему** при попытке "заимпортить что-то не то", а не гораздо позже)

**Reply by azinit** [reply]
> (3) Если у тиммейта появляются вопросы почему что то откуда то нельзя импортировать то вся логика описана в соотвествующем файле md (LEGACY.md или ARCHITECTURE.md) или вики страничке

Так вот мы как раз и хотим убежать от того когда "нужно написать и прочесть кучу документации", чтобы понимать что происходит в проекте

Хотелось бы, чтобы у разраба появлялись объяснения сразу на этапе линтования (это не значит что он не должен читать доку, но значит - что он не обязан блин помнить все эти простыни текста про то какая у нас здоровская и заморочистая архитектура0

**Reply by azinit** [reply]
> 5. кастомный снипет сразу мне создает экспортируемый FC c импортом React и строкой стейта

Вот и цена плохо настроенной автоматизации)))

*Но ладно, это уже ближе к кодстайлу, но все равно - хочется, чтобы разрабу, приходя в проект, нужно было сделать минимальное кол-во телодвижений, чтоб приносить пользу проетку*
